jax/scipy/sparse/linalg.py

```diff
@@ -54,42 +61,123 @@ def _identity(x):
 
   r0 = _sub(b, A(x0))
   p0 = z0 = M(r0)
-  gamma0 = _vdot_tree(r0, z0)
+  gamma0 = _vdot_tree_real(r0, z0)
   initial_value = (x0, r0, gamma0, p0, 0)
 
   x_final, *_ = lax.while_loop(cond_fun, body_fun, initial_value)
 
   return x_final
 
 
+def _bicgstab_solve(A, b, x0=None, *, maxiter, tol=1e-5, atol=0.0, M=_identity):
+
+  # tolerance handling uses the "non-legacy" behavior of scipy.sparse.linalg.bicgstab
+  bs = _vdot_tree_real(b, b)
+  atol2 = jnp.maximum(jnp.square(tol) * bs, jnp.square(atol))
+
+  # https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method#Preconditioned_BiCGSTAB
+
+  def cond_fun(value):
+    x, r, *_, k = value
+    rs = _vdot_tree_real(r, r)
+    return (rs > atol2) & (k < maxiter)
+
+  def body_fun(value):
+    x, r, rhat, alpha, omega, rho, p, q, k = value
+    # rho_ = _vdot_tree(_conj(rhat), r)
+    rho_ = _vdot_tree(rhat, r)
+    beta = rho_ / rho * alpha / omega
+    p_ = _add(r, _mul(beta, _sub(p, _mul(omega, q))))
+    phat = M(p_)
+    q_ = A(phat)
+    # alpha_ = rho_ / _vdot_tree(_conj(rhat), q_)
+    alpha_ = rho_ / _vdot_tree(rhat, q_)
+    s = _sub(r, _mul(alpha_, q_))
+    # TODO(?): stop early
+    #  It requires accessing cond_fun like this, but it's
+    #  not possible with jit...
+    #  if cond_fun((x, s, r0, alpha_, omega, rho_, p_, q_, k)):
+    #    x_ = _add(x, _mul(alpha_, phat))
+    #    return x_, s, rhat, alpha_, omega, rho_, p_, q, k
+    shat = M(s)
+    t = A(shat)
+    # omega_ = _vdot_tree(_conj(s), t) / _vdot_tree_real(t, t)
+    omega_ = _vdot_tree(s, t) / _vdot_tree_real(t, t)
+    x_ = _add(x, _add(_mul(alpha_, phat), _mul(omega_, shat)))
+    r_ = _sub(s, _mul(omega_, t))
+    return x_, r_, rhat, alpha_, omega_, rho_, p_, q_, k + 1
+
+  r0 = _sub(b, A(x0))
+  rho0 = alpha0 = omega0 = _vdot_tree(r0, r0)  # / _vdot_tree(r0, r0)
+  initial_value = (x0, r0, r0, alpha0, omega0, rho0, r0, r0, 0)
+
+  x_final, *_ = lax.while_loop(cond_fun, body_fun, initial_value)
+
+  return x_final
+
+
 def _shapes(pytree):
   return map(jnp.shape, tree_leaves(pytree))
 
 
+def isolve(_isolve, A, b, x0=None, *, tol=1e-5, atol=0.0,
+           maxiter=None, M=None):
+  if x0 is None:
+    x0 = tree_map(jnp.zeros_like, b)
+
+  b, x0 = device_put((b, x0))
+
+  if maxiter is None:
+    size = sum(bi.size for bi in tree_leaves(b))
+    maxiter = 10 * size  # copied from scipy
+
+  if M is None:
+    M = _identity
+
+  if tree_structure(x0) != tree_structure(b):
+    raise ValueError(
+        'x0 and b must have matching tree structure: '
+        f'{tree_structure(x0)} vs {tree_structure(b)}')
+
+  if _shapes(x0) != _shapes(b):
+    raise ValueError(
+        'arrays in x0 and b must have matching shapes: '
+        f'{_shapes(x0)} vs {_shapes(b)}')
+
+  isolve = partial(_isolve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
+
+  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  symmetric = all(map(real_valued, tree_leaves(b)))
```

Comment:
I think this condition is only true for `cg`, not `bicgstab`? I would expect arguments to `bicgstab` to essentially never be symmetric. As I understand it, if they were, there are better solvers to use (e.g., `cg`).

Reply:
I think if you have a matrix that is symmetric but NOT positive definite (e.g. pos semidefinite), this would still be useful, no? Don't know how often that would be the case in practice but it's certainly a possibility.

Reply:
OK, but either way it's definitely a different condition for `bicgstab` vs `cg`. All real values for the solution does not imply that it's symmetric.

So it seems like `isolve` should grow a `symmetric` argument, which could be set in a solver specific way. Potentially we could even expose a new `symmetric` argument on the public `bicgstab` function.

Reply:
Alright, and I also wonder if there is some utility in treating `isolve` as a public function or no? I can give it an underscore in the name if not!

Reply:
Yes, all the helper functions should definitely be marked as private.